[Package Signing] Add Certificates Validator

loic-sharma opened the pull request:
This Validator integrates with the validation Orchestrator to kick off one or more `ValidateCertificate` jobs for signed packages. The `PackageCertificatesValidator` will wait until the `ValidateCertificate` jobs completes before finalizing the overall validation.

Fixes https://github.com/NuGet/Engineering/issues/786.

dnfclas commented:
@loic-sharma,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

joelverhagen commented on tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidator.cs:
```diff
@@ -0,0 +1,299 @@
+
+namespace NuGet.Services.Validation.PackageSigning
+{
+    public class PackageCertificatesValidatorFacts
```
>PackageCertificatesValidatorFacts [](start = 17, length = 33)

Rename this file to match the class name.

joelverhagen commented on src/NuGet.Services.Validation.Orchestrator/PackageSigning/PackageSigningValidator.cs:
```diff
@@ -61,7 +61,7 @@ public class PackageSigningValidator : IValidator
                 // message has been duplicated.
                 _logger.LogError(
```
>LogError [](start = 24, length = 8)

Is this really an error? This feels more like a warning.

loic-sharma replied:
Sure. I'll move all other `LogError` to `LogWarning` for database concurrency issues.

joelverhagen commented on src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs:
```diff
@@ -0,0 +1,367 @@
+
+                InvalidatePackageSignatures(request, package, invalidSignatures);
+
+                return await TryUpdateValidationStatus(request, status, ValidationStatus.Failed);
```
>TryUpdateValidationStatus [](start = 29, length = 25)

`GetStatusAsync` is intended to be read-only (no concerns of concurrency, could be run on the secondary, etc etc). Can we make this read-only?

joelverhagen replied:
Oh and most importantly I was hoping this should be callable from admin panel without concerns of data changes.

---
In reply to: [143584031](https://github.com/NuGet/NuGet.Jobs/pull/225#discussion_r143584031) [](ancestors = 143584031)

loic-sharma replied:
Discussed this with Joel offline at great lengths. Ideally, the `GetStatusAsync` should be *completely* read-only. However, achieving this would complicate the pipeline a little. I personally do not think that the trade-off is worth it as of right now. If this becomes an issue, we can move the core of `GetStatusAsync` to its own job, and have that update the `ValidatorStatus` when all of the `CertificateValidation`s complete.

joelverhagen commented on src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs:
```diff
@@ -0,0 +1,367 @@
+            var signatures = await FindSignaturesAsync(request);
+            var invalidSignatures = signatures.Where(SignatureShouldBeInvalidated).ToList();
+
+            if (invalidSignatures.Any())
```
>invalidSignatures [](start = 16, length = 17)

This looks similar to code above in `GetStatusAsync` -- can we share the logic in a function?

loic-sharma replied:
Removed this logic from `GetStatusAsync`

joelverhagen commented on src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs:
```diff
@@ -0,0 +1,367 @@
+            return _validationContext
+                    .CertificateValidations
+                    .Where(v => v.ValidationId == request.ValidationId)
+                    .AllAsync(v => v.IsFinished());
```
I don't think `IsFinished` can be evaluated in SQL Server right? This either needs to be an expression or after the query is materialized.

loic-sharma replied:
Good catch. This wasn't caught by unit tests, but this did reproduce when I made a test console app for integration testing.

joelverhagen commented on src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs:
```diff
@@ -0,0 +1,367 @@
+        }
+
+        /// <summary>
+        /// Check whetehr all certificate validations for the given validation request are finished.
```
>whetehr [](start = 18, length = 7)

whether

joelverhagen commented on src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs:
```diff
@@ -0,0 +1,367 @@
+        }
+
+        /// <summary>
+        /// Enqueues certificate verifications and add <see cref="CertificateValidation"/> entities
```
>Enqueues [](start = 12, length = 8)

To have consistent voice of your other comments: "Enqueue" instead of "Enqueues"

xavierdecoster commented on src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs:
```diff
@@ -0,0 +1,367 @@
+        private IValidatorStateService _validatorStateService;
+        private ICertificateVerifier _certificateVerifier;
+        private TimeSpan _certificateValidationCacheSpan;
+        private ILogger<PackageCertificatesValidator> _logger;
```
All of the above fields can be made `readonly`.

loic-sharma commented:
Addressed all feedback, please review :)

agr commented on src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs:
```diff
@@ -0,0 +1,380 @@
+        /// </summary>
+        /// <param name="certificate">The certificate that may be revalidated.</param>
+        /// <returns>Whether the certificate should be revalidated.</returns>
+        private bool ShouldValidateCertificate(Certificate certificate)
```
`static`

scottbommarito commented on tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs:
```diff
@@ -0,0 +1,1098 @@
+
+            [Theory]
+            [MemberData(nameof(RevokedCertificatesFailsValiationOnlyIfRevokedBeforeSignatureTimeData))]
+            public async Task RevokedCertificatesFailsValiationOnlyIfRevokedBeforeSignatureTime(
```
`Valiation` -> `Validation`

loic-sharma replied:
Nice catch :)

loic-sharma commented on src/NuGet.Services.Validation.Orchestrator/PackageCertificates/CertificateVerifier.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace NuGet.Services.Validation.PackageCertificates
+{
+    public class CertificateVerifier : ICertificateVerifier
```
Name change sounds good. We have a generic service bus serialization class, but not an enqueuer class.

loic-sharma commented on src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs:
```diff
@@ -0,0 +1,380 @@
+            _certificateVerifier = certificateVerifier ?? throw new ArgumentNullException(nameof(certificateVerifier));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            _certificateValidationCacheSpan = certificateStatusCacheTime ?? TimeSpan.FromDays(1);
```
I suppose `TimeSpan.Zero` could be used if one intended to never cache certificates' statuses. But yes, a negative TimeSpan would be nonsensical :)

dtivel commented on tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs:
```diff
@@ -0,0 +1,936 @@
+        private const string PackageNormalizedVersion = "4.3.0.0-ALPHA+git";
+        private static readonly Guid ValidationId = new Guid("fb9c0bac-3d4d-4cc7-ac2d-b3940e15b94d");
+        private static readonly Guid OtherValidationId = new Guid("6593BD33-ABC0-4049-BDCF-915807F1D2B3");
+        private const string NupkgUrl = "https://example/nuget.versioning/4.3.0/package.nupkg";
```
Consider using [`unit.test`](https://en.wikipedia.org/wiki/.test) instead of `example`.

dtivel commented on tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs:
```diff
@@ -0,0 +1,936 @@
+                    new CertificateStatus?[] { CertificateStatus.Good, CertificateStatus.Unknown, CertificateStatus.Good },
+                };
+
+                // The validator should be resilient to no certificates. This should be an impossible state.
```
Unsigned packages will have `ValidationStatus.Succeeded` and no certificates.  See [here](https://github.com/NuGet/NuGet.Jobs/pull/225/files#diff-6d7330dc27fcee16ad09316d52d642e3R112).  How is it impossible?

loic-sharma replied:
This comment is confusing. This test ensures that **signed** packages are properly labelled by `GetStatusAsync` as succeeded/failed/incomplete based off the status of their certificate validations. A **signed** package should have at least 1 certificate, hence why this state should be "impossible". Note that this validator runs on packages whose signatures have passed the client's signature validations, as per the `ExtractAndValidatePackageSignature` job. Thus, it is reasonable for this validator to assume that a signature will have at least one certificate.

TLDR: The comment is misleading, I will improve it :)

dtivel commented on src/NuGet.Services.Validation.Orchestrator/NuGet.Services.Validation.Orchestrator.csproj:
```diff
@@ -48,6 +48,9 @@
     <Compile Include="IValidatorStateService.cs" />
     <Compile Include="Job.cs" />
+    <Compile Include="PackageCertificates\CertificateVerifier.cs" />
+    <Compile Include="PackageCertificates\ICertificateVerifier.cs" />
```
I think this project won't compile.  These files have been renamed, but the .csproj doesn't contain the new file names.

loic-sharma replied:
My bad. I've been doing a bunch of csproj hacks (replacing PackageReferences with ProjectReferences) to work around the current code signing issues, and I've been trying to avoid committing those hacks in.

dtivel commented on tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs:
```diff
@@ -0,0 +1,936 @@
+
+                yield return new object[]
+                {
+                    ValidationStatus.Incomplete,
```
I'm not sure if it's just an issue with this validation status or what, but "incomplete" implies non-finality.  However, if a certificate has been revoked, that's a final state for verification; there's nothing unfinished about the validation.

It seems like the validation status should be failed.

loic-sharma replied:
A validation will eventually fail if there are one or more revoked certificates - HOWEVER, the validator is designed to wait until ALL certificate validations have completed. If a certificate validation hasn't finished (aka a `null` `CertificateStatus`), the validator will say that the validation is `Incomplete`. The orchestrator will continue to ask the `CertificateValidator` for the latest status on the validations - and the `CertificateValidator` will eventually return `Failed` instead of `Incomplete`.

Why should the validator wait? A package may have multiple more than one invalid certificate. If the validator doesn't wait until all certificates are validated, it may miss invalid certificates and return an incorrect view on the package.

I will add a comment explaining this behavior :)

dtivel commented on tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs:
```diff
@@ -0,0 +1,936 @@
+            }
+
+            [Fact]
+            public async Task OnRevalidationAllNonRevokedCertificatesAreVerified()
```
I may have missed it, but do you have a test verifying that if a package has multiple signatures and one signature is with a revoked certificate, that revalidating non-revoked certificates won't mark the package as valid?

loic-sharma replied:
Interesting. I do not have any revalidation tests, I will see what additional corner cases I can come up with!

Side-note: once the `ValidateCertificate` job is running, I will add thorough integration testing between the `CertificatesValidator` and the `ValidateCertificate` job. This will be done in a separate pull request.

loic-sharma commented on src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs:
```diff
@@ -0,0 +1,333 @@
+
+            if (_certificateValidationCacheSpan < TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException("The certificate status cache time must be a positive value", nameof(certificateStatusCacheTime));
```
Weird - I could have sworn the parameter name came after the message in exceptions' constructors. I'll fix this :)

loic-sharma replied:
I found the source of my confusion. [`ArgumentException`](https://msdn.microsoft.com/en-us/library/sxykka64(v=vs.110).aspx) has the following signature:

```
public ArgumentException(string message, string paramName);
```

That's a pretty serious mistake by the .NET team 😢

loic-sharma commented on tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs:
```diff
@@ -0,0 +1,936 @@
+                        Status = certificateStatus,
+                        Certificate = certificate
+                    };
+
```
Nit: extra space

loic-sharma commented:
Updated to use latest version of the validation context, rebased to the new feature branch `dev-package-signing`, and addressed all feedback.

*NOTE*: The spec has changed! This pull request does not reflect the latest changes. I will implement the changes tomorrow.

loic-sharma commented on src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs:
```diff
@@ -0,0 +1,380 @@
+            IValidationEntitiesContext validationContext,
+            IValidatorStateService validatorStateService,
+            ICertificateVerifier certificateVerifier,
+            ILogger<PackageCertificatesValidator> logger,
```
Fixed

loic-sharma commented:
:bell: The validator is now feature complete. Please do one last round of review! :)
